lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -0,0 +1,162 @@
+"""
+pbench-clear-tools
+
+This script will remove tools that have been registered.  If no options are
+used, then all tools from the "default" tool group are removed.  Specifying
+a tool name and/or remote host will limit the scope of the removal.
+
+
+"""
+import sys
+import pathlib
+
+import click
+
+from pbench.agent import fs
+from pbench.agent.utils import setup_logging
+from pbench.cli.agent import CliContext, pass_cli_context
+from pbench.cli.agent.commands.tools.base import ToolCommand
+from pbench.cli.agent.options import common_options
+
+
+class ClearTools(ToolCommand):
+    def __init__(self, context):
+        super(ClearTools, self).__init__(context)
+
+        self.logger = setup_logging(logfile=self.pbench_log)
+
+    def execute(self):
+        """Clear registered tools"""
+        self.tools_group_dir = self.tool_group_dir(self.context.group)
+        if not self.tools_group_dir.exists():
+            return 1
+
+        try:
+            self.context.remote = self.context.remote.split(",")
+        except Exception:
+            # We were not given any remotes on the command line, build the list
+            # from the tools group directory.
+            self.context.remote = self.remote
+
+        for remote in self.context.remote:
+            tg_dir = self.tools_group_dir / remote
+            if not tg_dir.exists():
+                # Ignore any specified remote that does not exist.
+                self.logger.warn(
+                    f'The given remote host, "{remote}", is not a directory in {self.tools_group_dir}.'
+                )
+                continue
+
+            if self.context.name:
+                # Use the given tool name from the command line
+                self.context.name = [self.context.name]
+            else:
+                # Discovery all the tools registered for this remote
+                self.context.name = self.tools(tg_dir)
+
+            for name in self.context.name:
+                self._clear_tools(name, remote)
+
+            tool_files = [p.name for p in self.tools(tg_dir)]
+            if "__label__" in tool_files:
+                label = tg_dir / "__label__"
+                status = fs.safe_delete(label)
+                if status != 0:
+                    self.logger.error(f"Failed to remove label for remote {tg_dir}")
+
+            if self.is_empty(tg_dir):
+                self.logger.info(f'All tools removed from host, "{remote}"')
+                status = fs.safe_rmtree(tg_dir)
+                if status != 0:
+                    self.logger.error(f"Failed to remove remote directory {tg_dir}")
```

Comment:
For this usage, I'd expect the `ENOENT` should be "more than" just an error case, because the code flow here requires that they exist at the call. If this is the intent for all callers, then it should be formalized as part of the method contract and we should either throw on `ENOENT` (which I think would mean you could just remove the `except` clause) or possibly even an `assert` in the `except` clause so it doesn't escape.

Reply:
fixed